#source('./R/init.R') #Loading the external setting file

library(hwriter)
library(ggplot2)
library(grid)
library(hexbin)
numeDF <- function(x){
        as.numeric(x)
}
#
mfreqDF <- function(x){
        signif(max(x)/sum(x), digits=3)
}
#
countmatch <- function(x){
        length(x[!is.na(x)])
}

# 
#  argument
#
args <- commandArgs(trailingOnly = TRUE)
inR <- as.character(args[1])        #log file
# version <- "ref_map.pl version 1.37 started at 2017-02-09 11:30:33"
# dbname <- "TezukaRAD1_170209_radtags"
# batch_id <- 1
# cmd_stacks <- "ref_map.pl -T 12 -B TezukaRAD1_170209_radtags -b 1 -D Reference aligned population RAD-Tag samples -o ./stacks"
# sample_num <- 56
# totalread_num <- 191906430
# sample2readnum
#in_stacksout <-  as.character(args[2])
#in_title <- as.character(args[3])  #title
#in_enzyme1 <- as.character(args[4]) #reaction enzyme1
#in_enzyme2 <- as.character(args[5]) #reaction enzyme2
#in_polymerase  <- as.character(args[6]) #polymerase
#out_dir <- as.character(args[7])
#html_desc <- as.character(args[8])
#in_email <- as.character(args[8])  #email
#finish_file <- "finish"

#DEBUG
if(F){
  inR <- "sample/sample.R"   
  in_stacksout <- "sample/outputfile.tsv"
  in_title <- "Test"
  in_dname <- "dname"
  in_enzyme1 <- "AA1"
  in_enzyme2 <- "BB2"
  in_polymerase <- "CC3"
  out_dir <- "outxx2"
  finish_file <- "finish"
}

source(inR)

#denovo/ref
wr_rnum <- "Read number"
wr_rd   <- "Read"
wr_trnum<- "Total read number"
if (is.na(charmatch("denovo_map.pl",version))){
  wr_rnum <- "Primary alignment number"
  wr_rd   <- "Primary alignment"
  wr_trnum<- "Total primary alignment number" 
}
ggmain.size=14
gglab.size=13

#directory name (modified title)
in_dname <- gsub("[^a-zA-Z0-9_]","",in_title,perl=TRUE) 
if(nchar(in_dname) < 1){
  in_dname <- "radseq_report"
}

#data directory
repdate <- as.character(format(Sys.time(), "%Y-%m-%d-%H%M"))
reportID <- paste(in_dname, repdate, sep="_")
reportID_dir <- paste(out_dir,reportID,sep="/")
out_data_dir <- paste(reportID_dir,"data",sep="/");
out_html_dir <- paste(reportID_dir,"html",sep="/");
out_img_dir <- paste(out_html_dir,"img",sep="/");
dir.create(out_dir)
dir.create(reportID_dir)
dir.create(out_data_dir)
dir.create(out_html_dir)
dir.create(out_img_dir)
file.copy(html_desc,out_html_dir)
# directory image
# reportID
#  |-index.html
#  |-data
#  |  |-xxxx
#  |  |-xxxx
#  |  |-xxxx
#  |-html
#     |-xxxx
#     |-img
#        |-xxx

#read output file (ex. output.csv) generated by "export_sql.pl" stacks
data0 <- read.delim(in_stacksout, as.is=T, na.strings="")
r <- nrow(data0)
#id <- ncol(data0)-12
df <- r-sample_num

data <- head(data0, n=df) #read output.csv 
data0 <- NULL #
d1S <- subset(data, Num.SNPs==1 & Num.Alleles==2)
dwS <- subset(data, Num.SNPs>0)

depthsampleID <- 1:sample_num
stacksread <- 1:sample_num
meddepth <- 1:sample_num
maxdepth <- 1:sample_num

#setwd("StacksData")
catd <- data[1]
nr <- nrow(data)
ed <- ncol(data)
namelist <- colnames(data[13:ed])
nL <- length(namelist)
for(j in 1:nL){
	h <- namelist[j]
	tmpdir <- paste(out_data_dir,h,sep="/");
	ds <- subset(data[h], is.na(data[h]) == FALSE)
	colnames(ds) <- "depth"
	dD <- transform(ds, m.al.freq=NA)
	mds <- apply(as.matrix(ds), 1, as.character)
	splitmds <- strsplit(mds, "/")
	resultnume <- lapply(splitmds, numeDF)
	resultsum <- lapply(resultnume, sum)
	resultfreq <- lapply(resultnume, mfreqDF)
	dD[1] <- unlist(resultsum)
	dD["m.al.freq"] <- unlist(resultfreq)
	dir.create(tmpdir)
	#setwd(h)
	write.table (dD, file = paste(tmpdir,"alldepth.txt",sep="/"), sep = "\t", quote = FALSE, row.names = FALSE)
	depthsampleID[j] <- h
	dDu <- dD$depth
	stacksread[j] <- sum(dDu)
	meddepth[j] <- median(dDu)
	maxdepth[j] <- max(dDu)
	#setwd("../")
	catd[h] <- rep(0,nr)
	ins <- rownames(dD)
	catd[ins,h] <- dD[ins,1]
}
#setwd("../")
depthsummary <- data.frame(depthsampleID, stacksread, meddepth, maxdepth)
write.table (depthsummary, file = paste(out_data_dir,"depthsummary.txt",sep="/"), sep = "\t", quote = FALSE, row.names = FALSE)

#setwd("StacksData")
d2 <- dwS
ed <- ncol(d2)
namelist <- colnames(d2[13:ed])
nL <- length(namelist)
for(j in 1:nL){
	h <- namelist[j]
	tmpdir <- paste(out_data_dir,h,sep="/");
	d2s <- subset(d2[h], is.na(d2[h]) == FALSE)
	colnames(d2s) <- "depth"
	d2D <- transform(d2s, m.al.freq=NA)
	md2s <- apply(as.matrix(d2s), 1, as.character)
	splitmd2s <- strsplit(md2s, "/")
	resultnume <- lapply(splitmd2s, numeDF)
	resultsum <- lapply(resultnume, sum)
	resultfreq <- lapply(resultnume, mfreqDF)
	d2D[1] <- unlist(resultsum)
	d2D["m.al.freq"] <- unlist(resultfreq)
	#setwd(h)
	write.table (d2D, file = paste(tmpdir,"wSNPdepth.txt",sep="/"), sep = "\t", quote = FALSE, row.names = FALSE)
	#setwd("../")
}
#setwd("../")

nr <- nrow(data)
nl <- ncol(data)
l <- data[13:nl]
ml <- as.matrix(l)
connumber <- 1:nr
MatchingSamples <- apply(ml, 1, countmatch)
mr <- MatchingSamples*100/(nl-12)
MSratio <- signif(mr, digits=2)
result.d <- data.frame(connumber, MatchingSamples, MSratio)
write.table (result.d, file = paste(out_data_dir,"allmatch.txt",sep="/"), sep = "\t", quote = FALSE, row.names = FALSE)

nr <- nrow(d1S)
nl <- ncol(d1S)
l <- d1S[13:nl]
ml <- as.matrix(l)
connumber <- 1:nr
MatchingSamples <- apply(ml, 1, countmatch)
mr <- MatchingSamples*100/(nl-12)
MSratio <- signif(mr, digits=2)
result.d1S <- data.frame(connumber, MatchingSamples, MSratio)
write.table (result.d1S, file = paste(out_data_dir,"AlwSNPmatch.txt",sep="/"), sep = "\t", quote = FALSE, row.names = FALSE)

nr <- nrow(dwS)
nl <- ncol(dwS)
l <- dwS[13:nl]
ml <- as.matrix(l)
connumber <- 1:nr
MatchingSamples <- apply(ml, 1, countmatch)
mr <- MatchingSamples*100/(nl-12)
MSratio <- signif(mr, digits=2)
result.dwS <- data.frame(connumber, MatchingSamples, MSratio)
write.table (result.dwS, file = paste(out_data_dir,"wSNPmatch.txt",sep="/"), sep = "\t", quote = FALSE, row.names = FALSE)

#clean
d1S <- NULL
dwS <- NULL
data <- NULL

#setwd(wd0)

Ltop <- hwrite("Library Summary (General)")

SampleSummary <- data.frame(read.number=as.integer(sample2readnum))
rownames(SampleSummary) <- names(sample2readnum)
totalreadnum <- sum(SampleSummary$read.number)
readnumps.me <- median(SampleSummary$read.number)
rn1q <- floor(quantile(SampleSummary$read.number, p=0.25))
rn3q <- floor(quantile(SampleSummary$read.number, p=0.75))
readnumps.q <- paste(rn1q, "-", rn3q)
readnumps.max <- max(SampleSummary$read.number)
readnumps.min <- min(SampleSummary$read.number)
#setwd(reportID)
SampleSummary$log10read <- log10(sample2readnum+1)
ReadNumberHist <- ggplot(SampleSummary, aes(x = log10read))+geom_histogram(binwidth=0.5, fill="seagreen4")+ xlim(0,8)+ labs(title=paste("Histogram of ",tolower(wr_rnum),"\nper sample",sep=""), x=bquote(paste(Log[10]," ",.(tolower(wr_rnum)),sep="")), y="Number of samples")+ theme(plot.title = element_text(size = ggmain.size),axis.title.x = element_text(size=gglab.size),axis.title.y = element_text(size=gglab.size))
ggsave(file = paste(out_img_dir,"ReadNumberHist.png",sep="/"), plot = ReadNumberHist, dpi = 100, width = 4, height = 4)
# averageQVHist <- ggplot(SampleSummary, aes(x = average.QV))+geom_histogram(binwidth=1, fill="seagreen4")+ xlim(0,41)+ labs(title="Histogram of average QV", x="average QV", y="Frequency")+ theme(plot.title = element_text(size = rel(1.1)))
# ggsave(file = "averageQVHist.png", plot = averageQVHist, dpi = 100, width = 4, height = 4)
# ReadvsQV <- ggplot(SampleSummary, aes(x = log10read, y = average.QV))+geom_point(shape=1, size=3)+ xlim(0,8)+ ylim(0,41)+ labs(title="Read number vs average QV", x="log10 read number", y="average QV")+ theme(plot.title = element_text(size = rel(1.1)))
# ggsave(file = "ReadvsQV.png", plot = ReadvsQV, dpi = 100, width = 4, height = 4)
sgg1 <- hwrite(c(tolower(wr_trnum), totalreadnum), col.width=c('200px','125px'), border=0, cellspacing=0)
sgg0 <- hwrite(" ", br=TRUE)
sgg2 <- hwrite(c(paste(wr_rnum," per sample",sep=""), " "), col.width=c('200px','150px'), border=0, cellspacing=0)
sgg3 <- hwrite(c(" ", "median", " ", readnumps.me), col.width=c('20px','90px','15px','150px'), border=0, cellspacing=0, bgcolor=c("",'#aaffaa',"",""))
sgg4 <- hwrite(c(" ", "1q-3q range", " ", readnumps.q), col.width=c('20px','90px','15px','150px'), border=0, cellspacing=0, bgcolor=c("",'#aaffaa',"",""))
sgg5 <- hwrite(c(" ", "max", " ", readnumps.max), col.width=c('20px','90px','15px','150px'), border=0, cellspacing=0, bgcolor=c("",'#aaffaa',"",""))
sgg6 <- hwrite(c(" ", "min", " ", readnumps.min), col.width=c('20px','90px','15px','150px'), border=0, cellspacing=0, bgcolor=c("",'#aaffaa',"",""))
sggL1L <- hwrite(c(sgg1, sgg0, sgg2, sgg3, sgg4, sgg5, sgg6), dim=c(7,1), col.names=FALSE, border=0, cellspacing=0, width='325px')
sggL1R <- hwriteImage("./img/ReadNumberHist.png", width='325px')
#sggL2L <- hwriteImage("./img/averageQVHist.png", width='325px')
#sggL2R <- hwriteImage("./img/ReadvsQV.png", width='325px')

guideI <- hwrite("index", link='../index.html', center=TRUE)
guideLR <- hwrite("Library Summary RAD", link='LibrarySummaryRAD.html', center=TRUE)
guideLG <- hwrite("Library Summary General", center=TRUE)
guide1 <- hwrite(c(guideI, guideLG, guideLR), col.names=FALSE, border=0, cellspacing=0, col.width=c('150px','250px','250px'))

p <- openPage(paste(out_html_dir,"LibrarySummaryGeneral.html",sep="/"), css='TD{font-size:11pt}')
hwrite(guide1, p, border=0, cellspacing=0, col.width='650px')
hwrite(Ltop, p, heading=4)
#hwrite(days, p)
hwrite("", p, br=TRUE)
hwrite(c(sggL1L, sggL1R), p, border=0, cellspacing=0)
#hwrite(c(sggL2L, sggL2R), p, border=0, cellspacing=0)
hwrite("", p, br=TRUE)
hwrite(c("memo", " "), p, col.width=c('50px','650px'), height='100px', cellspacing=0)
closePage(p, splash=FALSE)
# setwd("../")
rm(sgg0,sgg1,sgg2,sgg3,sgg4,sgg5,sgg6,sggL1L,sggL1R,guide1,guideI,guideLG,guideLR)


#setwd(reportID)
#dir.create(paste(reportID_dir,"img",sep="/"))
#setwd("../")
#Osamplelist <- read.delim("samplelist.txt", as.is=T, na.strings="", header=F)
OL <- length(sample2readnum)
mOsamplelist <- rep(NA, OL)
for(i in 1:OL){
	h0 <- sub('\\.[^.]*', "", names(sample2readnum)[i])
	mOsamplelist[i] <- sub('-', "\\.", h0)
}

tmpx <- NA
tmpy <- NA

depthIDlist <- depthsummary$depthsampleID

i <- 1
	h <- mOsamplelist[i]
	read.number.s <- SampleSummary[i, "read.number"]
	#XX#aveQV.s <- SampleSummary[i, "average.QV"]
	guideI <- hwrite("index", link='../index.html', center=TRUE)
	guideLG <- hwrite("Library Summary General", link='LibrarySummaryGeneral.html', center=TRUE)
	guideLR <- hwrite("Library Summary RAD", link='LibrarySummaryRAD.html', center=TRUE)
	NL <- paste(mOsamplelist[i+1], ".html", sep="")
	NN <- paste(mOsamplelist[i+1], ">>", sep=" ")
	guideN <- hwrite(NN, link=NL)
	guide1 <- hwrite(c(guideI, guideLG, guideLR), col.names=FALSE, border=0, cellspacing=0, col.width=c('150px','250px','250px'))
	guide2 <- hwrite(c("", guideN), col.names=FALSE, border=0, cellspacing=0, col.width=c('325px','325px'), style=c('text-align:left','text-align:right'))
	SN <- paste("Sample Name : ", h, sep="")
	SStop <- hwrite(SN, style='font-weight: bold; font-size:110%')
	SSgtop <- hwrite("Sample Summary (General)", style='font-weight: bold; font-size:105%')
	SSgL1 <- hwrite(c(tolower(wr_rnum), read.number.s), col.width=c('180px','145px'), border=0, cellspacing=0)
	#SSgL2 <- hwrite(c("average QV", aveQV.s), col.width=c('180px','145px'), border=0, cellspacing=0)
	SSg0 <- hwrite(" ", br=TRUE)
	SSgL3 <- hwrite(c(SSgL1), dim=c(2,1), col.names=FALSE, border=0, cellspacing=0)
	SSgL <- hwrite(c(SStop, SSg0, SSgtop, SSg0, SSg0, SSgL3), dim=c(6,1), col.names=FALSE, border=0, cellspacing=0)
	#XX#QVname0 <- paste("QVplot/", Osamplelist[i,], "_qv.png", sep="")
	#XX#QVname1 <- paste(out_img_dir, "/", Osamplelist[i,], "_qv.png", sep="")
	#XX#file.rename(QVname0, QVname1)
	#XX#QVname <- paste("img/", Osamplelist[i,], "_qv.png", sep="")
	#XX#SSgR <- hwriteImage(QVname, width='325px')
	depthdetect <- match(h, depthIDlist, nomatch=0)
	if (depthdetect > 0) {
		SSrtop <- hwrite("Sample Summary (RAD-Seq)", style='font-weight: bold; font-size:105%')
		dsdata <- depthsummary[depthsummary$depthsampleID == h,]
		SSr1 <- hwrite(c(paste(tolower(wr_rnum)," used in stacks",sep=""), dsdata$stacksread), col.width=c('210px','115px'), border=0, cellspacing=0)
		SSr2 <- hwrite(c("median depth", dsdata$meddepth), col.width=c('210px','115px'), border=0, cellspacing=0)
		SSr3 <- hwrite(c("max depth", dsdata$maxdepth), col.width=c('210px','115px'), border=0, cellspacing=0)
		SSr13 <- hwrite(c(SSr1, SSr2, SSr3), dim=c(3,1), col.names=FALSE, border=0, cellspacing=0, br=TRUE)
		SSrL <- hwrite(c(SSrtop, SSg0, SSg0, SSr13), dim=c(4,1), col.names=FALSE, border=0, cellspacing=0, br=TRUE)
		dpass1 <- paste(out_data_dir, h, "alldepth.txt", sep="/")
		dpass2 <- paste(out_data_dir, h, "wSNPdepth.txt", sep="/")
		alldepth <- read.delim(dpass1, as.is=T, na.strings="")
		alld <- nrow(alldepth["depth"])
		alld3 <- nrow(subset(alldepth["depth"], alldepth["depth"]>=3))
		alld10 <- nrow(subset(alldepth["depth"], alldepth["depth"]>=10))
		alld30 <- nrow(subset(alldepth["depth"], alldepth["depth"]>=30))
		alld100 <- nrow(subset(alldepth["depth"], alldepth["depth"]>=100))
		wSNPdepth <- read.delim(dpass2, as.is=T, na.strings="")
		wSNPd <- nrow(wSNPdepth["depth"])
		wSNPd3 <- nrow(subset(wSNPdepth["depth"], wSNPdepth["depth"]>=3))
		wSNPd10 <- nrow(subset(wSNPdepth["depth"], wSNPdepth["depth"]>=10))
		wSNPd30 <- nrow(subset(wSNPdepth["depth"], wSNPdepth["depth"]>=30))
		wSNPd100 <- nrow(subset(wSNPdepth["depth"], wSNPdepth["depth"]>=100))
		dL1 <- hwrite(c("Number of loci","all","with SNP"), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor='#aaffaa', border=0, cellspacing=0)
		dL2 <- hwrite(c("all",alld,wSNPd), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		dL3 <- hwrite(c("depth >= 3",alld3,wSNPd3), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		dL4 <- hwrite(c("depth >= 10",alld10,wSNPd10), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		dL5 <- hwrite(c("depth >= 30",alld30,wSNPd30), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		dL6 <- hwrite(c("depth >= 100",alld100,wSNPd100), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		tabledata <- hwrite(c(dL1, dL2, dL3, dL4, dL5, dL6), dim=c(6,1), col.names=FALSE, border=0, cellspacing=0)
		tableall <- hwrite(c(SSrL,tabledata), dim=c(2,1), col.names=FALSE, border=0, cellspacing=0)
		#setwd(reportID)
		png1 <- paste(out_img_dir,"/", h, "_depth3.png", sep="")
		png1_link <- paste("img/", h, "_depth3.png", sep="")
		png2 <- paste(out_img_dir,"/", h, "_freq500.png", sep="")
		png2_link <- paste("img/", h, "_freq500.png", sep="")
		imagep_link <- paste("img/", h, "_imageplot.png", sep="")
		
		png(png2, width = 400, height = 400)
		par(mar=c(4, 3.5, 3, 0.3), mgp=c(2.5, 0.7, 0))
		depth3 <- subset(alldepth["depth"], alldepth["depth"]>=3)
		depth3$upper100 <- replace(depth3[[1]], depth3[[1]]>100, 100)
		hist(depth3$upper100, breaks=seq(0,105,5), right=FALSE, ylim=c(0,500), col="seagreen4", xaxt="n", main="", xlab="", ylab="", border=NA, las=1, cex.axis=1.1)
		axis(side=1,at=seq(0,100,25), cex.axis=1.1)
		title(xlab="Depth", ylab="Number of loci", main="Histogram of depth (Number of loci < 500)", cex.lab=1.4, cex.main=1.4)
		dev.off()
		
		alldepth01 <- subset(alldepth, alldepth[[2]]<1)
		y100 <- replace(alldepth01[[1]], alldepth01[[1]]>100, 100)
		tmpx <- c(tmpx, alldepth01[[2]])
		tmpy <- c(tmpy, y100)
		histR <- hwriteImage(png1_link, width='325px')
		histL <- hwriteImage(png2_link, width='325px')
		plotR <- hwriteImage(imagep_link, height='325px')
		filename <- paste(out_html_dir, "/", h, ".html", sep="")
		p <- openPage(filename, css='TD{font-size:11pt}')
		hwrite(guide1, p, border=0, cellspacing=0, col.width='650px')
		hwrite(guide2, p, border=0, cellspacing=0, col.width='650px')
		hwrite("", p, br=TRUE)
		hwrite(c(SSgL), p, border=0, cellspacing=0, style='vertical-align:top')
		hwrite(c(tableall,histR), p, border=0, cellspacing=0, col.width=c('325px','325px'), style='vertical-align:top')
		hwrite(c(histL, plotR), p, border=0, cellspacing=0)
		closePage(p, splash=FALSE)
		#setwd("../")
	} else {
		#setwd(reportID)
		filename <- paste(out_html_dir, "/", h, ".html", sep="")
		p <- openPage(filename, css='TD{font-size:11pt}')
		hwrite(guide1, p, border=0, cellspacing=0, col.width='650px')
		hwrite(guide2, p, border=0, cellspacing=0, col.width='650px')
		hwrite("", p, br=TRUE)
		hwrite(c(SSgL), p, border=0, cellspacing=0, style='vertical-align:top')
		closePage(p, splash=FALSE)
		#setwd("../")
	}

for(i in 2:(OL-1)){
	h <- mOsamplelist[i]
	read.number.s <- SampleSummary[i, "read.number"]
	#XX#aveQV.s <- SampleSummary[i, "average.QV"]
	guideI <- hwrite("index", link='../index.html', center=TRUE)
	guideLG <- hwrite("Library Summary General", link='LibrarySummaryGeneral.html', center=TRUE)
	guideLR <- hwrite("Library Summary RAD", link='LibrarySummaryRAD.html', center=TRUE)
	PL <- paste(mOsamplelist[i-1], ".html", sep="")
	PN <- paste("<<", mOsamplelist[i-1], sep=" ")
	guideP <- hwrite(PN, link=PL)
	NL <- paste(mOsamplelist[i+1], ".html", sep="")
	NN <- paste(mOsamplelist[i+1], ">>", sep=" ")
	guideN <- hwrite(NN, link=NL)
	guide1 <- hwrite(c(guideI, guideLG, guideLR), col.names=FALSE, border=0, cellspacing=0, col.width=c('150px','250px','250px'))
	guide2 <- hwrite(c(guideP, guideN), col.names=FALSE, border=0, cellspacing=0, col.width=c('325px','325px'), style=c('text-align:left','text-align:right'))
	SN <- paste("Sample Name : ", h, sep="")
	SStop <- hwrite(SN, style='font-weight: bold; font-size:110%')
	SSgtop <- hwrite("Sample Summary (General)", style='font-weight: bold; font-size:105%')
	SSgL1 <- hwrite(c(tolower(wr_rnum), read.number.s), col.width=c('180px','145px'), border=0, cellspacing=0)
	#XX#SSgL2 <- hwrite(c("average QV", aveQV.s), col.width=c('180px','145px'), border=0, cellspacing=0)
	SSg0 <- hwrite(" ", br=TRUE)
	SSgL3 <- hwrite(c(SSgL1), dim=c(2,1), col.names=FALSE, border=0, cellspacing=0)
	SSgL <- hwrite(c(SStop, SSg0, SSgtop, SSg0, SSg0, SSgL3), dim=c(6,1), col.names=FALSE, border=0, cellspacing=0)
	#XX#QVname0 <- paste("QVplot/", Osamplelist[i,], "_qv.png", sep="")
	#XX#QVname1 <- paste(reportID, "/img/", Osamplelist[i,], "_qv.png", sep="")
	#XX#file.rename(QVname0, QVname1)
	#XX#QVname <- paste("img/", Osamplelist[i,], "_qv.png", sep="")
	#XX#SSgR <- hwriteImage(QVname, width='325px')
	depthdetect <- match(h, depthIDlist, nomatch=0)
	if (depthdetect > 0) {
		SSrtop <- hwrite("Sample Summary (RAD-Seq)", style='font-weight: bold; font-size:105%')
		dsdata <- depthsummary[depthsummary$depthsampleID == h,]
		SSr1 <- hwrite(c(paste(tolower(wr_rnum)," used in stacks",sep=""), dsdata$stacksread), col.width=c('210px','115px'), border=0, cellspacing=0)
		SSr2 <- hwrite(c("median depth", dsdata$meddepth), col.width=c('210px','115px'), border=0, cellspacing=0)
		SSr3 <- hwrite(c("max depth", dsdata$maxdepth), col.width=c('210px','115px'), border=0, cellspacing=0)
		SSr13 <- hwrite(c(SSr1, SSr2, SSr3), dim=c(3,1), col.names=FALSE, border=0, cellspacing=0, br=TRUE)
		SSrL <- hwrite(c(SSrtop, SSg0, SSg0, SSr13), dim=c(4,1), col.names=FALSE, border=0, cellspacing=0, br=TRUE)
		dpass1 <- paste(out_data_dir, h, "alldepth.txt", sep="/")
		dpass2 <- paste(out_data_dir, h, "wSNPdepth.txt", sep="/")
		alldepth <- read.delim(dpass1, as.is=T, na.strings="")
		alld <- nrow(alldepth["depth"])
		alld3 <- nrow(subset(alldepth["depth"], alldepth["depth"]>=3))
		alld10 <- nrow(subset(alldepth["depth"], alldepth["depth"]>=10))
		alld30 <- nrow(subset(alldepth["depth"], alldepth["depth"]>=30))
		alld100 <- nrow(subset(alldepth["depth"], alldepth["depth"]>=100))
		wSNPdepth <- read.delim(dpass2, as.is=T, na.strings="")
		wSNPd <- nrow(wSNPdepth["depth"])
		wSNPd3 <- nrow(subset(wSNPdepth["depth"], wSNPdepth["depth"]>=3))
		wSNPd10 <- nrow(subset(wSNPdepth["depth"], wSNPdepth["depth"]>=10))
		wSNPd30 <- nrow(subset(wSNPdepth["depth"], wSNPdepth["depth"]>=30))
		wSNPd100 <- nrow(subset(wSNPdepth["depth"], wSNPdepth["depth"]>=100))
		dL1 <- hwrite(c("contig number","all","with SNP"), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor='#aaffaa', border=0, cellspacing=0)
		dL2 <- hwrite(c("all",alld,wSNPd), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		dL3 <- hwrite(c("depth >= 3",alld3,wSNPd3), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		dL4 <- hwrite(c("depth >= 10",alld10,wSNPd10), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		dL5 <- hwrite(c("depth >= 30",alld30,wSNPd30), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		dL6 <- hwrite(c("depth >= 100",alld100,wSNPd100), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		tabledata <- hwrite(c(dL1, dL2, dL3, dL4, dL5, dL6), dim=c(6,1), col.names=FALSE, border=0, cellspacing=0)
		tableall <- hwrite(c(SSrL,tabledata), dim=c(2,1), col.names=FALSE, border=0, cellspacing=0)
		#setwd(reportID)
		png1 <- paste(out_img_dir,"/", h, "_depth3.png", sep="")
		png1_link <- paste("img/", h, "_depth3.png", sep="")
		png2 <- paste(out_img_dir,"/", h, "_freq500.png", sep="")
		png2_link <- paste("img/", h, "_freq500.png", sep="")
		imagep_link <- paste("img/", h, "_imageplot.png", sep="")
		
		png(png2, width = 400, height = 400)
		par(mar=c(4, 3.5, 3, 0.3), mgp=c(2.5, 0.7, 0))
		depth3 <- subset(alldepth["depth"], alldepth["depth"]>=3)
		depth3$upper100 <- replace(depth3[[1]], depth3[[1]]>100, 100)
		hist(depth3$upper100, breaks=seq(0,105,5), right=FALSE, ylim=c(0,500), col="seagreen4", xaxt="n", main="", xlab="", ylab="", border=NA, las=1, cex.axis=1.1)
		axis(side=1,at=seq(0,100,25), cex.axis=1.1)
		title(xlab="Depth", ylab="Number of loci", main="Histogram of depth (Number of loci < 500)", cex.lab=1.4, cex.main=1.4)
		dev.off()
		alldepth01 <- subset(alldepth, alldepth[[2]]<1)
		y100 <- replace(alldepth01[[1]], alldepth01[[1]]>100, 100)
		tmpx <- c(tmpx, alldepth01[[2]])
		tmpy <- c(tmpy, y100)
		histR <- hwriteImage(png1_link, width='325px')
		histL <- hwriteImage(png2_link, width='325px')
		plotR <- hwriteImage(imagep_link, height='325px')
		filename <- paste(out_html_dir, "/", h, ".html", sep="")
		p <- openPage(filename, css='TD{font-size:11pt}')
		hwrite(guide1, p, border=0, cellspacing=0, col.width='650px')
		hwrite(guide2, p, border=0, cellspacing=0, col.width='650px')
		hwrite("", p, br=TRUE)
		hwrite(c(SSgL), p, border=0, cellspacing=0, style='vertical-align:top')
		hwrite(c(tableall,histR), p, border=0, cellspacing=0, col.width=c('325px','325px'), style='vertical-align:top')
		hwrite(c(histL, plotR), p, border=0, cellspacing=0)
		closePage(p, splash=FALSE)
		#setwd("../")
 	} else {
		#setwd(reportID)
		filename <- paste(out_html_dir, "/", h, ".html", sep="")
		p <- openPage(filename, css='TD{font-size:11pt}')
		hwrite(guide1, p, border=0, cellspacing=0, col.width='650px')
		hwrite(guide2, p, border=0, cellspacing=0, col.width='650px')
		hwrite("", p, br=TRUE)
		hwrite(c(SSgL), p, border=0, cellspacing=0, style='vertical-align:top')
		closePage(p, splash=FALSE)
		#setwd("../")
	}
}

i <- OL
	h <- mOsamplelist[i]
	read.number.s <- SampleSummary[i, "read.number"]
	#XX#aveQV.s <- SampleSummary[i, "average.QV"]
	guideI <- hwrite("index", link='../index.html', center=TRUE)
	guideLG <- hwrite("Library Summary General", link='LibrarySummaryGeneral.html', center=TRUE)
	guideLR <- hwrite("Library Summary RAD", link='LibrarySummaryRAD.html', center=TRUE)
	PL <- paste(mOsamplelist[i-1], ".html", sep="")
	PN <- paste("<<", mOsamplelist[i-1], sep=" ")
	guideP <- hwrite(PN, link=PL)
	guide1 <- hwrite(c(guideI, guideLG, guideLR), col.names=FALSE, border=0, cellspacing=0, col.width=c('150px','250px','250px'))
	guide2 <- hwrite(c(guideP, ""), col.names=FALSE, border=0, cellspacing=0, col.width=c('325px','325px'), style=c('text-align:left','text-align:right'))
	SN <- paste("Sample Name : ", h, sep="")
	SStop <- hwrite(SN, style='font-weight: bold; font-size:110%')
	SSgtop <- hwrite("Sample Summary (General)", style='font-weight: bold; font-size:105%')
	SSgL1 <- hwrite(c(tolower(wr_rnum), read.number.s), col.width=c('180px','145px'), border=0, cellspacing=0)
	#XX#SSgL2 <- hwrite(c("average QV", aveQV.s), col.width=c('180px','145px'), border=0, cellspacing=0)
	SSg0 <- hwrite(" ", br=TRUE)
	SSgL3 <- hwrite(c(SSgL1), dim=c(2,1), col.names=FALSE, border=0, cellspacing=0)
	SSgL <- hwrite(c(SStop, SSg0, SSgtop, SSg0, SSg0, SSgL3), dim=c(6,1), col.names=FALSE, border=0, cellspacing=0)
	#XX#QVname0 <- paste("QVplot/", Osamplelist[i,], "_qv.png", sep="")
	#XX#QVname1 <- paste(reportID, "/img/", Osamplelist[i,], "_qv.png", sep="")
	#XX#file.rename(QVname0, QVname1)
	#XX#QVname <- paste("img/", Osamplelist[i,], "_qv.png", sep="")
	#XX#SSgR <- hwriteImage(QVname, width='325px')
	depthdetect <- match(h, depthIDlist, nomatch=0)
	if (depthdetect > 0) {
		SSrtop <- hwrite("Sample Summary (RAD-Seq)", style='font-weight: bold; font-size:105%')
		dsdata <- depthsummary[depthsummary$depthsampleID == h,]
		SSr1 <- hwrite(c(paste(tolower(wr_rnum)," used in stacks",sep=""), dsdata$stacksread), col.width=c('210px','115px'), border=0, cellspacing=0)
		SSr2 <- hwrite(c("median depth", dsdata$meddepth), col.width=c('210px','115px'), border=0, cellspacing=0)
		SSr3 <- hwrite(c("max depth", dsdata$maxdepth), col.width=c('210px','115px'), border=0, cellspacing=0)
		SSr13 <- hwrite(c(SSr1, SSr2, SSr3), dim=c(3,1), col.names=FALSE, border=0, cellspacing=0, br=TRUE)
		SSrL <- hwrite(c(SSrtop, SSg0, SSg0, SSr13), dim=c(4,1), col.names=FALSE, border=0, cellspacing=0, br=TRUE)
		dpass1 <- paste(out_data_dir, h, "alldepth.txt", sep="/")
		dpass2 <- paste(out_data_dir, h, "wSNPdepth.txt", sep="/")
		alldepth <- read.delim(dpass1, as.is=T, na.strings="")
		alld <- nrow(alldepth["depth"])
		alld3 <- nrow(subset(alldepth["depth"], alldepth["depth"]>=3))
		alld10 <- nrow(subset(alldepth["depth"], alldepth["depth"]>=10))
		alld30 <- nrow(subset(alldepth["depth"], alldepth["depth"]>=30))
		alld100 <- nrow(subset(alldepth["depth"], alldepth["depth"]>=100))
		wSNPdepth <- read.delim(dpass2, as.is=T, na.strings="")
		wSNPd <- nrow(wSNPdepth["depth"])
		wSNPd3 <- nrow(subset(wSNPdepth["depth"], wSNPdepth["depth"]>=3))
		wSNPd10 <- nrow(subset(wSNPdepth["depth"], wSNPdepth["depth"]>=10))
		wSNPd30 <- nrow(subset(wSNPdepth["depth"], wSNPdepth["depth"]>=30))
		wSNPd100 <- nrow(subset(wSNPdepth["depth"], wSNPdepth["depth"]>=100))
		dL1 <- hwrite(c("contig number","all","with SNP"), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor='#aaffaa', border=0, cellspacing=0)
		dL2 <- hwrite(c("all",alld,wSNPd), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		dL3 <- hwrite(c("depth >= 3",alld3,wSNPd3), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		dL4 <- hwrite(c("depth >= 10",alld10,wSNPd10), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		dL5 <- hwrite(c("depth >= 30",alld30,wSNPd30), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		dL6 <- hwrite(c("depth >= 100",alld100,wSNPd100), col.width=c('120px','80px','90px'), col.style=c('text-align:left','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
		tabledata <- hwrite(c(dL1, dL2, dL3, dL4, dL5, dL6), dim=c(6,1), col.names=FALSE, border=0, cellspacing=0)
		tableall <- hwrite(c(SSrL,tabledata), dim=c(2,1), col.names=FALSE, border=0, cellspacing=0)
		#setwd(reportID)
		png1 <- paste(out_img_dir,"/", h, "_depth3.png", sep="")
		png1_link <- paste("img/", h, "_depth3.png", sep="")
		png2 <- paste(out_img_dir,"/", h, "_freq500.png", sep="")
		png2_link <- paste("img/", h, "_freq500.png", sep="")
		imagep_link <- paste("img/", h, "_imageplot.png", sep="")
		
		png(png2, width = 400, height = 400)
		par(mar=c(4, 3.5, 3, 0.3), mgp=c(2.5, 0.7, 0))
		depth3 <- subset(alldepth["depth"], alldepth["depth"]>=3)
		depth3$upper100 <- replace(depth3[[1]], depth3[[1]]>100, 100)
		hist(depth3$upper100, breaks=seq(0,105,5), right=FALSE, ylim=c(0,500), col="seagreen4", xaxt="n", main="", xlab="", ylab="", border=NA, las=1, cex.axis=1.1)
		axis(side=1,at=seq(0,100,25), cex.axis=1.1)
		title(xlab="Depth", ylab="Number of loci", main="Histogram of depth (Number of loci < 500)", cex.lab=1.4, cex.main=1.4)
		dev.off()
		alldepth01 <- subset(alldepth, alldepth[[2]]<1)
		y100 <- replace(alldepth01[[1]], alldepth01[[1]]>100, 100)
		tmpx <- c(tmpx, alldepth01[[2]])
		tmpy <- c(tmpy, y100)
		histR <- hwriteImage(png1_link, width='325px')
		histL <- hwriteImage(png2_link, width='325px')
		plotR <- hwriteImage(imagep_link, height='325px')
		filename <- paste(out_html_dir, "/", h, ".html", sep="")
		p <- openPage(filename, css='TD{font-size:11pt}')
		hwrite(guide1, p, border=0, cellspacing=0, col.width='650px')
		hwrite(guide2, p, border=0, cellspacing=0, col.width='650px')
		hwrite("", p, br=TRUE)
		hwrite(c(SSgL), p, border=0, cellspacing=0, style='vertical-align:top')
		hwrite(c(tableall,histR), p, border=0, cellspacing=0, col.width=c('325px','325px'), style='vertical-align:top')
		hwrite(c(histL, plotR), p, border=0, cellspacing=0)
		closePage(p, splash=FALSE)
		#setwd("../")
 	} else {
		#setwd(reportID)
		filename <- paste(out_html_dir, "/", h, ".html", sep="")
		p <- openPage(filename, css='TD{font-size:11pt}')
		hwrite(guide1, p, border=0, cellspacing=0, col.width='650px')
		hwrite(guide2, p, border=0, cellspacing=0, col.width='650px')
		hwrite("", p, br=TRUE)
		hwrite(c(SSgL), p, border=0, cellspacing=0, style='vertical-align:top')
		closePage(p, splash=FALSE)
		#setwd("../")
	}

for(i in 1:OL){
	h <- mOsamplelist[i]
	depthdetect <- match(h, depthIDlist, nomatch=0)
	if (depthdetect > 0) {
		dpass1 <- paste(out_data_dir, h, "alldepth.txt", sep="/")
		alldepth <- read.delim(dpass1, as.is=T, na.strings="")
		depth3 <- subset(alldepth["depth"], alldepth["depth"]>=3)
		depth3$upper100 <- replace(depth3[[1]], depth3[[1]]>100, 100)
		#setwd(reportID)
		
		# ooshima
		if(length(depth3$upper100) > 0){
			png1 <- paste(out_img_dir, "/", h, "_depth3.png", sep="")
			d3 <- ggplot(depth3, aes(x=upper100))+geom_histogram(binwidth=5, fill="seagreen4")+ scale_x_continuous(limits=c(0,105), labels=c("0", "25", "50", "75", ">100"))+labs(title="Histogram of depth (depth >= 3)", x="Depth", y="Number of loci")+ theme(plot.title = element_text(size = ggmain.size),axis.title.x = element_text(size=gglab.size),axis.title.y = element_text(size=gglab.size))
			ggsave(file = png1, plot = d3, dpi = 1000, width = 4, height = 4)
		}
		
		alldepth01 <- subset(alldepth, alldepth[[2]]<1)
		y100 <- replace(alldepth01[[1]], alldepth01[[1]]>100, 100)
		if(length(y100) != 0 ){
			gb2 <- ggplot(alldepth01, aes(x = alldepth01[[2]], y = y100))
			imagep <- paste(out_img_dir, "/", h, "_imageplot.png", sep="")
			ip <- gb2+ geom_hex()+ scale_x_continuous(limits = c(0, 1), breaks=seq(0,1,0.2))+ scale_fill_continuous(trans = "log10")+ scale_y_continuous(limits = c(0, 100), labels = c("0", "25", "50", "75", ">100"))+labs(title="Major allele depth vs total depth", x="rate of major allele in the locus", y="Depth of each locus")+ theme(legend.margin = unit(0, "cm"))
			ggsave(file = imagep, plot = ip, dpi = 100, width = 4.5, height = 4)
	 	}
		#setwd("../")
	}
}


allmatch <- read.delim(paste(out_data_dir,"allmatch.txt",sep="/"), as.is=T, na.strings="")
AlwSNPmatch <- read.delim(paste(out_data_dir,"AlwSNPmatch.txt",sep="/"), as.is=T, na.strings="")
wSNPmatch <- read.delim(paste(out_data_dir,"wSNPmatch.txt",sep="/"), as.is=T, na.strings="")

all.all <- length(allmatch$MatchingSamples)
all.50 <- nrow(allmatch[allmatch$MSratio>50, ])
all.70 <- nrow(allmatch[allmatch$MSratio>70, ])
all.90 <- nrow(allmatch[allmatch$MSratio>90, ])
all.100 <- nrow(allmatch[allmatch$MSratio==100, ])

Alw.all <- length(AlwSNPmatch$MatchingSamples)
Alw.50 <- nrow(AlwSNPmatch[AlwSNPmatch$MSratio>50, ])
Alw.70 <- nrow(AlwSNPmatch[AlwSNPmatch$MSratio>70, ])
Alw.90 <- nrow(AlwSNPmatch[AlwSNPmatch$MSratio>90, ])
Alw.100 <- nrow(AlwSNPmatch[AlwSNPmatch$MSratio==100, ])

wSNP.all <- length(wSNPmatch$MatchingSamples)
wSNP.50 <- nrow(wSNPmatch[wSNPmatch$MSratio>50, ])
wSNP.70 <- nrow(wSNPmatch[wSNPmatch$MSratio>70, ])
wSNP.90 <- nrow(wSNPmatch[wSNPmatch$MSratio>90, ])
wSNP.100 <- nrow(wSNPmatch[wSNPmatch$MSratio==100, ])

matchx <-1:19
matchy <- 1:19 
for(i in 1:19){
	p <- 10+(i-1)*5
	matchx[i] <- p
	matchy[i] <- nrow(allmatch[allmatch$MSratio>=p, ])
}
pallmatch <- data.frame(matching.samples = matchx, contig = matchy)
for(i in 1:19){
	p <- 10+(i-1)*5
	matchx[i] <- p
	matchy[i] <- nrow(AlwSNPmatch[AlwSNPmatch$MSratio>=p, ])
}
pAlwSNPmatch <- data.frame(matching.samples = matchx, contig = matchy)
for(i in 1:19){
	p <- 10+(i-1)*5
	matchx[i] <- p
	matchy[i] <- nrow(wSNPmatch[wSNPmatch$MSratio>=p, ])
}
pwSNPmatch <- data.frame(matching.samples = matchx, contig = matchy)

#setwd(reportID)
png(paste(out_img_dir,"matchingvscontig.png",sep="/"), width = 400, height = 400)
par(mar=c(4, 3.5, 2, 1), mgp=c(2.2, 0.7, 0))
plot(contig~matching.samples, pallmatch, type="l", xaxt="n", ylim=c(0, max(pallmatch$contig)), col="steelblue2", xlab="Matching sample (%)",ylab="Contig", main="Contig number", cex.main=1.4, cex.lab=1.4, cex.axis=1.2)
axis(side=1,at=seq(10,100,10), cex.axis=1.2)
panel.first = grid(NA, NULL, lty = 2)
polygon(c(pallmatch$matching.samples, rev(pallmatch$matching.samples)), c(rep(0, length(pallmatch$contig)), rev(pallmatch$contig)), col="steelblue2", border = NA)
lines(contig~matching.samples, pwSNPmatch, type="l", xaxt="n", col="orange")
polygon(c(pwSNPmatch$matching.samples, rev(pwSNPmatch$matching.samples)), c(rep(0, length(pwSNPmatch$contig)), rev(pwSNPmatch$contig)), col="orange", border = NA)
lines(contig~matching.samples, pAlwSNPmatch, type="l", xaxt="n", col="yellow")
polygon(c(pAlwSNPmatch$matching.samples, rev(pAlwSNPmatch$matching.samples)), c(rep(0, length(pAlwSNPmatch$contig)), rev(pAlwSNPmatch$contig)), col="yellow", border = NA)
legend(43, max(pallmatch$contig), paste(c("all","contigs with (mutiple)SNP(s)","contigs with 1SNP 2 alleles" )), col=c("steelblue2","orange","yellow"), pch=15, cex=1.2, bg="white")
dev.off()
#setwd("../")


depthsummary <- read.delim(paste(out_data_dir,"depthsummary.txt",sep="/"), as.is=T, na.strings="")
#setwd(reportID)
depthsummary$lreadnumstacks <- log10(depthsummary$stacksread+1)
readnumstacks <- ggplot(depthsummary, aes(x = lreadnumstacks))+geom_histogram(binwidth=0.5, fill="seagreen4")+ xlim(0,8)+ labs(title=paste("Histogram of ",tolower(wr_rnum),"\nused in stacks",sep=""), x=bquote(paste(Log[10]," ",.(tolower(wr_rnum)),sep="")), y="Number of samples")+ theme(plot.title = element_text(size = ggmain.size),axis.title.x = element_text(size=gglab.size),axis.title.y = element_text(size=gglab.size))
ggsave(file = paste(out_img_dir,"readnumstacks.png",sep="/"), plot = readnumstacks, dpi = 100, width = 4, height = 4)
meddepthstacks <- ggplot(depthsummary, aes(x = meddepth))+geom_histogram(binwidth=1, fill="seagreen4")+ labs(title="Median depth of samples", x="Median depth", y="Number of samples")+ theme(plot.title = element_text(size = ggmain.size),axis.title.x = element_text(size=gglab.size),axis.title.y = element_text(size=gglab.size))
ggsave(file = paste(out_img_dir,"mediandepthHist.png",sep="/"), plot = meddepthstacks, dpi = 100, width = 4, height = 4)
maxdepthstacks <- ggplot(depthsummary, aes(x = maxdepth))+geom_histogram(binwidth=50, fill="seagreen4")+ labs(title="Max depth of samples", x="Max depth", y="Number of samples")+ theme(plot.title = element_text(size = ggmain.size),axis.title.x = element_text(size=gglab.size),axis.title.y = element_text(size=gglab.size))
ggsave(file = paste(out_img_dir,"maxdepthHist.png",sep="/"), plot = maxdepthstacks, dpi = 100, width = 4, height = 4)
#setwd("../")

nct <- ncol(catd)
rct <- nrow(catd)
catdl <- catd[1]
forn <- floor(rct/20000) 
for(i in 0:(forn-1)){
	start <- 1+(i*20000)
	end <- 20000*(i+1)
	catdtmp <- catd[start:end, 2:nct]
	m.catdtmp <- data.matrix(catdtmp)
	catdl[start:end, "average.depth"] <- rowSums(m.catdtmp)/ncol(m.catdtmp)
	catdl[start:end, "total.depth"] <- rowSums(m.catdtmp)
	#print(i)
}
catdtmp <- catd[(end+1):rct, 2:nct]
m.catdtmp <- data.matrix(catdtmp)
catdl[(end+1):rct, "average.depth"] <- rowSums(m.catdtmp)/ncol(m.catdtmp)
catdl[(end+1):rct, "total.depth"] <- rowSums(m.catdtmp)
#dim(catdl)
#print(rct)
write.table (catdl, file = paste(out_data_dir,"depthsummary.catalog.txt",sep="/"), sep = "\t", quote = FALSE, row.names = FALSE)

depthlist <- matrix(rep(NA, 6*4), nrow=6)
colnames(depthlist) <- c("average depth", "loci", paste("number of used ",tolower(wr_rd),sep=""), paste("percentage of used ",tolower(wr_rd)," (%)",sep=""))
j <- 1
for(i in c(0,1,3,10,30,100)){
	contigs <- subset(catdl, catdl$average.depth > i)
	ad <- paste(">", i, sep="")
	depthlist[j, "average depth"] <- ad
	depthlist[j, "loci"] <- nrow(contigs)
	depthlist[j, paste("number of used ",tolower(wr_rd),sep="")] <- sum(contigs$total.depth)
	depthlist[j, paste("percentage of used ",tolower(wr_rd)," (%)",sep="")] <- floor(sum(contigs$total.depth)*100/sum(catdl$total.depth))
	j <- j+1
}

#setwd(reportID)
catdl$laverage <- log10(catdl$average.depth)
adepth <- ggplot(catdl, aes(x = laverage))+geom_histogram(binwidth=0.2, fill="seagreen4")+ xlim(-3,5)+ labs(title="Average depth of locus", x=bquote(paste(Log[10]," (average depth)",sep="")), y="Number of loci")+ theme(plot.title = element_text(size = ggmain.size),axis.title.x = element_text(size=gglab.size),axis.title.y = element_text(size=gglab.size))
ggsave(file = paste(out_img_dir,"averagedepth.png",sep="/"), plot = adepth, dpi = 100, width = 4, height = 4)
png(paste(out_img_dir,"averagedepth3000.png",sep="/"), width = 400, height = 400)
par(mar=c(4, 3.5, 3, 0.3), mgp=c(2.5, 0.7, 0))
hist(catdl$laverage, breaks=seq(-4,6,0.2), right=FALSE, ylim=c(0,3000), col="seagreen4", xaxt="n", main="", xlab="", ylab="", border=NA, las=1, cex.axis=1.1)
axis(side=1,at=seq(-4,6,1), cex.axis=1.1)
title(xlab=bquote(paste(Log[10]," (average depth)",sep="")), ylab="Number of loci", main="Average depth of locus (frequency < 3000)", cex.lab=1.4, cex.main=1.4)
dev.off()

allsamplesdepth <- data.frame(depth=tmpx, y100=tmpy)
gb2 <- ggplot(allsamplesdepth, aes(x = depth, y = y100))
ip <- gb2+ geom_hex()+ scale_x_continuous(limits = c(0, 1), breaks=seq(0,1,0.2))+ scale_fill_continuous(trans = "log10")+ scale_y_continuous(limits = c(0, 100), labels = c("0", "25", "50", "75", ">100"))+labs(title="Major allele depth vs total depth", x="rate of major allele in the locus", y="Depth of each locus")+ theme(legend.margin = unit(0, "cm"))
ggsave(file = paste(out_img_dir,"allimageplot.png",sep="/"), plot = ip, dpi = 100, width = 4.5, height = 4)
#setwd("../")

#enzyme1 <- in_enzyme1
#enzyme2 <- in_enzyme2
#polym <- in_polymerase
#pcrcycle <- 20 ###
#stacksp <- "denovo_map.pl &nbsp;-m 3 -M 2 -n 0 -T 15 -t" 
#under1000 <- read.delim("Under1000Sample.txt", as.is=T, na.strings="")
#under1000 <- sample2readnum[which(sample2readnum < 1000)]
#write.table(under1000,file=paste(out_data_dir,"Under1000Sample.txt",sep="/"),col.names=F,quote=F)
#remsamples <- length(under1000)

Rtop <- hwrite("Library Summary (RAD-Seq)")
#RL1L <- hwrite(c("preparation method", prepmethod), col.width=c('200px','150px'), border=0, cellspacing=0)
#RL1R <- hwrite(c("polymerase", polym), col.width=c('200px','150px'), border=0, cellspacing=0)
#RL2L <- hwrite(c("restriction enzyme 1", enzyme1), col.width=c('200px','150px'), border=0, cellspacing=0)
#RL2R <- hwrite(c("PCR cycle", pcrcycle), col.width=c('200px','150px'), border=0, cellspacing=0)
#RL3L <- hwrite(c("restriction enzyme 2", enzyme2), col.width=c('200px','150px'), border=0, cellspacing=0)
#RL3R <- hwrite(c(" ", " "), col.width=c('200px','150px'), border=0, cellspacing=0)
#RL1 <- hwrite(c(RL1R), col.names=FALSE, border=0, cellspacing=0)
#RL2 <- hwrite(c(RL2L), col.names=FALSE, border=0, cellspacing=0)
#RL3 <- hwrite(c(RL3L, RL3R), col.names=FALSE, border=0, cellspacing=0)
#methods <- hwrite(c(RL1, RL2, RL3), dim=c(3,1), col.names=FALSE, border=0, cellspacing=0)

#remlink <- hwrite(remsamples, link='../Under1000Sample.txt')
#spL1L <- hwrite(c("samples removed from stacks", remlink), col.width=c('220px','80px'), border=0, cellspacing=0)
spL1R <- hwrite(c("database name", dbname), col.width=c('120px','220px'), border=0, cellspacing=0)
spL1 <- hwrite(c(spL1R), col.names=FALSE, border=0, cellspacing=0)
stacksnumber <- nrow(depthsummary)
spL2L <- hwrite(c("number of samples in stacks", stacksnumber), col.width=c('220px','80px'), border=0, cellspacing=0)
spL2R <- hwrite(c("batch id", batch_id), col.width=c('120px','220px'), border=0, cellspacing=0)
spL2 <- hwrite(c(spL2L, spL2R), col.names=FALSE, border=0, cellspacing=0)
spn <- hwrite(c(spL1, spL2), dim=c(2,1), col.names=FALSE, border=0, cellspacing=0)
spL3 <- hwrite(c("stacks parameters", cmd_stacks), col.width=c('150px','400px'), border=0, cellspacing=0)

cL1 <- hwrite(c("contigs","all","1SNP 2 alleles","with (mutiple)SNP(s)"), col.width=c('180px','90px','120px','100px'), col.style=c('text-align:left','text-align:right','text-align:right','text-align:right'), bgcolor='#aaffaa', border=0, cellspacing=0)
cL2 <- hwrite(c("all",all.all,Alw.all,wSNP.all), col.width=c('180px','90px','120px','100px'), col.style=c('text-align:left','text-align:right','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
cL3 <- hwrite(c("matching sample > 50%",all.50,Alw.50,wSNP.50), col.width=c('180px','90px','120px','100px'), col.style=c('text-align:left','text-align:right','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
cL4 <- hwrite(c("matching sample > 70%",all.70,Alw.70,wSNP.70), col.width=c('180px','90px','120px','100px'), col.style=c('text-align:left','text-align:right','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
cL5 <- hwrite(c("matching sample > 90%",all.90,Alw.90,wSNP.90), col.width=c('180px','90px','120px','100px'), col.style=c('text-align:left','text-align:right','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
cL6 <- hwrite(c("matching sample = 100%",all.100,Alw.100,wSNP.100), col.width=c('180px','90px','120px','100px'), col.style=c('text-align:left','text-align:right','text-align:right','text-align:right'), bgcolor=c('#aaffaa',"","",""), border=0, cellspacing=0)
con <- hwrite(c(cL1, cL2, cL3, cL4, cL5, cL6), dim=c(6,1), col.names=FALSE, border=0, cellspacing=0)

#setwd(reportID)
pwddir="./img/"
Rf1L <- hwriteImage(paste(pwddir,"matchingvscontig.png",sep="/"), width='325px')
Rf1R <- hwriteImage(paste(pwddir,"readnumstacks.png",sep="/"), width='325px')
Rf2L <- hwriteImage(paste(pwddir,"mediandepthHist.png",sep="/"), width='325px')
Rf2R <- hwriteImage(paste(pwddir,"maxdepthHist.png",sep="/"), width='325px')
P21 <- hwriteImage(paste(pwddir,"averagedepth.png",sep="/"), width='325px')
P22 <- hwriteImage(paste(pwddir,"averagedepth3000.png",sep="/"), width='325px')
P23 <- hwriteImage(paste(pwddir,"allimageplot.png",sep="/"), width='360px')

guideI <- hwrite("index", link='../index.html', center=TRUE)
guideLR <- hwrite("Library Summary RAD", center=TRUE)
guideLG <- hwrite("Library Summary General", link='LibrarySummaryGeneral.html', center=TRUE)
guide1 <- hwrite(c(guideI, guideLG, guideLR), col.names=FALSE, border=0, cellspacing=0, col.width=c('150px','250px','250px'))

p <- openPage(paste(out_html_dir,"LibrarySummaryRAD.html",sep="/"), css='TD{font-size:11pt}')
hwrite(guide1, p, border=0, cellspacing=0, col.width='650px')
hwrite(Rtop, p, heading=4)
#hwrite(methods, p)
hwrite("", p, br=TRUE)
hwrite(spn, p)
hwrite(spL3, p)
hwrite("", p, br=TRUE)
hwrite("", p, br=TRUE)
hwrite(con, p)
hwrite("", p, br=TRUE)
hwrite(c(Rf1L, Rf1R), p, border=0, cellspacing=0)
hwrite(c(Rf2L, Rf2R), p, border=0, cellspacing=0)
hwrite("", p, br=TRUE)
hwrite("", p, br=TRUE)
hwrite(depthlist, p, row.bgcolor='#ffffaa', table.style=list('text-align:center'), border=0, cellspacing=0, cellpadding=5, table.frame='void', table.rules='cols', table.hspace=5)
hwrite("", p, br=TRUE)
hwrite("", p, br=TRUE)
hwrite(c(P21, P22), p, border=0, cellspacing=0)
hwrite("", p, br=TRUE)
hwrite("", p, br=TRUE)
hwrite(P23, p, border=0, cellspacing=0)
closePage(p, splash=FALSE)

#setwd("../")

#XX#dname <- basename(wd0)
#XX#wd0を利用しないようにする
sentense <- paste("Library Report of ", in_title, sep="")
itop <- hwrite(sentense, style='font-weight: bold; font-size:105%', br=TRUE)
LSGlink <- paste("./html","LibrarySummaryGeneral.html", sep="/")
LSG <- hwrite("Library Summary (General)", link=LSGlink)
LSGd <- hwrite(c("General information about the library", "This section contains information such as the date of the Stack analysis, number of samples, and histograms of read number and quality value."), dim=c(2,1), col.names=FALSE, border=0)
L1 <- hwrite(c(LSG,LSGd), border=0, cellspacing=0, col.width=c('200px','400px'), style='vertical-align:top')
LSRlink <- paste("./html","LibrarySummaryRAD.html", sep="/")
LSR <- hwrite("Library Summary (RAD)", link=LSRlink)
LSRd <- hwrite(c("Information about the RAD-Seq library", "This section contains information such as Stacks parameters used, the number of loci in each percentage of matching samples, and histograms of read number, depth, and other metrics."), dim=c(2,1), col.names=FALSE, border=0)
L2 <- hwrite(c(LSR,LSRd), border=0, cellspacing=0, col.width=c('200px','400px'), style='vertical-align:top')
SS1stlink <- paste("./html/", mOsamplelist[1], ".html", sep="")
SSlastlink <- paste("./html/", mOsamplelist[OL], ".html", sep="")
SS1st <- hwrite(">> Summary of 1st Sample", link=SS1stlink)
SSlast <- hwrite(">> Summary of last Sample", link=SSlastlink)
SSd <- hwrite(c("Information about each sample", "General: read number, mean quality value, and their relationship.","RAD-Seq: Stacks read number, histograms of depths, the relationship between the number of loci and depth, and the relationship between depth of major alleles and total depth of alleles at each locus."), dim=c(3,1), col.names=FALSE, border=0)
SSR <- hwrite(c(SSd, SSg0, SS1st, SSlast), dim=c(4,1), col.names=FALSE, border=0, cellspacing=0, col.width='400px')
L3 <- hwrite(c("Sample Summary",SSR), border=0, cellspacing=0, col.width=c('200px','400px'), style='vertical-align:top')
content <- hwrite("##### Contents #####", style='text-align:center')
list <- hwrite(c(SSg0,SSg0,L1,SSg0,SSg0,L2,SSg0,SSg0,L3), dim=c(9,1), border=0, cellspacing=0, center=TRUE)
explink <- paste("./html","figexp_e.txt", sep="/")
figexp <- hwrite("Description of each file", link=explink)
total <- hwrite(c(SSg0,itop,SSg0,SSg0,content,list,SSg0,SSg0,SSg0,figexp), dim=c(10,1), border=0, cellspacing=0, col.width='650px', style='text-align:center')

sampletable <- data.frame(Sample_name = mOsamplelist)
sampletable$num <- as.numeric(rownames(sampletable))
sampletable[wr_rnum] <- SampleSummary$read.number
usedinstacks <- depthsummary[1:2]
colnames(usedinstacks) <- c("Sample_name", paste(wr_rnum," used in stacks",sep=""))
sampletable <- merge(sampletable, usedinstacks, by="Sample_name", all=T, sort=F)
sampletable <- sampletable[order(sampletable$num),]
sampletable$num <- NULL
sampletable[paste("Percentage of used ",tolower(wr_rd)," (%)",sep="")] <- floor(sampletable[3]*100/sampletable[2])
samplelink <- paste("./html/", mOsamplelist, ".html", sep="")

p <- openPage(paste(reportID_dir,"index.html",sep="/"), css='TD{font-size:11pt}', charset = "UTF-8")
hwrite(total, p, border=0, cellspacing=0)
hwrite("", p, br=TRUE)
hwrite("", p, br=TRUE)
hwrite(sampletable, p, row.bgcolor='#ffffaa', col.link=list(Sample_name=samplelink), table.style=list('text-align:center'), border=0, cellspacing=0, cellpadding=5, table.frame='void', table.rules='cols', table.hspace=5, row.names=FALSE)
closePage(p, splash=FALSE)

#処理完了後にfinishへステータス変更
file.rename(inR, finish_file)
